service/storage_service.cc

```diff
@@ -1088,11 +1082,24 @@ void storage_service::handle_state_normal(inet_address endpoint) {
     if (_gossiper.uses_host_id(endpoint)) {
         auto host_id = _gossiper.get_host_id(endpoint);
         auto existing = _token_metadata.get_endpoint_for_host_id(host_id);
+        auto replace_addr = db().local().get_replace_address();
         if (db().local().is_replacing() &&
-            db().local().get_replace_address() &&
-                _gossiper.get_endpoint_state_for_endpoint_ptr(db().local().get_replace_address().value())  &&
-            (host_id == _gossiper.get_host_id(db().local().get_replace_address().value()))) {
-            slogger.warn("Not updating token metadata for {} because I am replacing it", endpoint);
+            replace_addr &&
+            _gossiper.get_endpoint_state_for_endpoint_ptr(*replace_addr) &&
+            host_id == _gossiper.get_host_id(*replace_addr)) {
```

Comment:
lines 1089 and 1088 are continuation lines, they should be indented (preferably 2 tabs)

Reply:
This commit mixes refactoring changes with logic changes. It should be split.

Also, now you're using the `*` operator on `std::optional` instead of `.value()`, which changes the semantics in case `replace_addr` was `nullopt` from throwing an error to undefined behavior.

`replace_addr` is probably not `nullopt`, but I'd suggest to check it with `assert`/`on_internal_error`, or keep using `value()` as before.

Reply:
Before:

db().local().is_replacing() &&
db().local().get_replace_address() && gossiper.get_endpoint_state_for_endpoint_ptr(db().local().get_replace_address().value())

After:

auto replace_addr = db().local().get_replace_address();
db().local().is_replacing() && replace_addr &&
_gossiper.get_endpoint_state_for_endpoint_ptr(*replace_addr)

The get_replace_address() is guaranteed to have a value, both before and after.

replace_addr is *guarnteed* to be not nullopt, not probably.

Are you suggesting to split the change that 

```
auto replace_addr = db().local().get_replace_address();
```

to a different patch? What is the point for such split?

I am calling auto replace_addr = db().local().get_replace_address() to avoid use  db().local().get_replace_address()  again and again in the new code which uses it multiple times.

Reply:
> replace_addr is guarnteed to be not nullopt, not probably

I meant that there "probably" isn't a bug which makes it nullopt :)

> to a different patch? What is the point for such split?

Just for review to be easier, but it's just a suggestion, I don't require it.

> I am calling auto replace_addr = db().local().get_replace_address() to avoid use db().local().get_replace_address() again and again in the new code which uses it multiple times.

I know, but you've also changed `X.value()` to `*X`, which might change the semantics.

Reply:
X.value() to *X, nothing changes if the X is guaranteed to be nullopt. Calling either  X.value() or *X is fine. *X is preferred because it avoids the unnecessary check. 

I am marking this as resolved. If you think we really need to change anything in the code. We can reopen.